From ee860546fec06446a3dcdbbdda636a7953858ab5 Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Mon, 18 Aug 2025 17:17:57 +0200
Subject: [PATCH 01/12] gluon-core: rename device_uses_11a to device_uses_5ghz

this is a more appropriate naming as the function checks if a radio uses the 5g band
---
 .../luasrc/lib/gluon/config-mode/wizard/0250-outdoor.lua        | 2 +-
 package/gluon-core/luasrc/lib/gluon/upgrade/180-outdoors        | 2 +-
 package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua        | 2 +-
 .../luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua    | 2 +-
 4 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/package/gluon-config-mode-outdoor/luasrc/lib/gluon/config-mode/wizard/0250-outdoor.lua b/package/gluon-config-mode-outdoor/luasrc/lib/gluon/config-mode/wizard/0250-outdoor.lua
index 947fa0b4ac..49a6a3b373 100644
--- a/package/gluon-config-mode-outdoor/luasrc/lib/gluon/config-mode/wizard/0250-outdoor.lua
+++ b/package/gluon-config-mode-outdoor/luasrc/lib/gluon/config-mode/wizard/0250-outdoor.lua
@@ -2,7 +2,7 @@ return function(form, uci)
 	local platform = require 'gluon.platform'
 	local wireless = require 'gluon.wireless'
 
-	if not (platform.is_outdoor_device() and wireless.device_uses_11a(uci)) then
+	if not (platform.is_outdoor_device() and wireless.device_uses_5ghz(uci)) then
 		-- only visible on wizard for outdoor devices
 		return
 	end
diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/180-outdoors b/package/gluon-core/luasrc/lib/gluon/upgrade/180-outdoors
index 59704bffc5..b75d3ce494 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/180-outdoors
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/180-outdoors
@@ -23,7 +23,7 @@ if sysconfig.gluon_version then
 	outdoor = false
 elseif config == 'preset' then
 	-- enable outdoor mode through presets on new installs
-	outdoor = platform.is_outdoor_device() and wireless.device_uses_11a(uci)
+	outdoor = platform.is_outdoor_device() and wireless.device_uses_5ghz(uci)
 else
 	-- enable/disable outdoor mode unconditionally on new installs
 	outdoor = config
diff --git a/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua b/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua
index 5845bca7a7..5ded4f5f50 100644
--- a/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua
+++ b/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua
@@ -122,7 +122,7 @@ function M.device_uses_wlan(uci)
 	return ret
 end
 
-function M.device_uses_11a(uci)
+function M.device_uses_5ghz(uci)
 	local ret = false
 
 	uci:foreach('wireless', 'wifi-device', function(radio)
diff --git a/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua b/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
index 3f0169418a..92cda8efe0 100644
--- a/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
+++ b/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
@@ -130,7 +130,7 @@ uci:foreach('wireless', 'wifi-device', function(config)
 end)
 
 
-if wireless.device_uses_11a(uci) and not wireless.preserve_channels(uci) then
+if wireless.device_uses_5ghz(uci) and not wireless.preserve_channels(uci) then
 	local r = f:section(Section, translate("Outdoor Installation"), translate(
 		"Configuring the node for outdoor use tunes the 5 GHz radio to a frequency "
 		.. "and transmission power that conforms with the local regulatory requirements. "

From 1173e4b0abd963af3bb4a97867cead2b3d790d6b Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Wed, 20 Aug 2025 01:08:00 +0200
Subject: [PATCH 02/12] gluon-core: improve band counting

using band2g instead of band24 allows to retrieve the actual band name from the key.
Add a helper function to increment the correct counter
---
 .../luasrc/lib/gluon/upgrade/200-wireless       | 17 +++++++++++------
 1 file changed, 11 insertions(+), 6 deletions(-)

diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
index a4ca20219b..904141f619 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
@@ -12,19 +12,24 @@ local uci = require('simple-uci').cursor()
 if not sysconfig.gluon_version then
 	uci:delete_all('wireless', 'wifi-iface')
 
+	local radio_band_count = {band2g=0, band5g=0}
+
+	local function count_band(band)
+		radio_band_count[band] = radio_band_count[band] + 1
+	end
+
 	-- First count all radios with a fixed frequency band.
 	-- This is needed to distribute devices which have radios
 	-- capable of operating in the 2.4 GHz and 5 GHz band need
 	-- to be distributed evenly.
-	local radio_band_count = {band24=0, band5=0}
 	wireless.foreach_radio(uci, function(radio)
 		local hwmodes = iwinfo.nl80211.hwmodelist(wireless.find_phy(radio))
 		if hwmodes.g and not (hwmodes.a or hwmodes.ac) then
 			-- 2.4 GHz
-			radio_band_count["band24"] = radio_band_count["band24"] + 1
+			count_band("band2g")
 		elseif (hwmodes.a or hwmodes.ac) and not hwmodes.g then
 			-- 5 GHz
-			radio_band_count["band5"] = radio_band_count["band5"] + 1
+			count_band("band5g")
 		end
 	end)
 
@@ -35,13 +40,13 @@ if not sysconfig.gluon_version then
 		local hwmodes = iwinfo.nl80211.hwmodelist(wireless.find_phy(radio))
 		if (hwmodes.a or hwmodes.ac) and hwmodes.g then
 			-- Dualband radio
-			if radio_band_count["band24"] <= radio_band_count["band5"] then
+			if radio_band_count["band2g"] <= radio_band_count["band5g"] then
 				-- Assign radio to 2.4GHz band
-				radio_band_count["band24"] = radio_band_count["band24"] + 1
+				count_band("band2g")
 				uci:set('wireless', radio_name, 'band', '2g')
 			else
 				-- Assign radio to 5GHz band
-				radio_band_count["band5"] = radio_band_count["band5"] + 1
+				count_band("band5g")
 				uci:set('wireless', radio_name, 'band', '5g')
 			end
 		end

From 7a697e155b016310232790aff901f3a03c8b7f7c Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Wed, 20 Aug 2025 00:04:08 +0200
Subject: [PATCH 03/12] gluon-core: add initialization script of gluon wireless
 roles

migrate existing wireless configuration to the new role settings.
Existing roles are migrated to gluon.wireless section if the radio band is not configured yet.

This also checks to disable mesh if the device is an outdoor device.
---
 .../lib/gluon/upgrade/022-wireless-roles      | 92 +++++++++++++++++++
 1 file changed, 92 insertions(+)
 create mode 100755 package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles

diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles b/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles
new file mode 100755
index 0000000000..c62cd71f19
--- /dev/null
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles
@@ -0,0 +1,92 @@
+#!/usr/bin/lua
+
+local site = require 'gluon.site'
+local uci = require('simple-uci').cursor()
+local util = require 'gluon.util'
+local wireless = require 'gluon.wireless'
+
+-- Create wifi roles if they do not exist yet
+local wifi_roles = {
+	band2g = {'client'},
+	band5g = {'client'},
+}
+
+-- enable mesh if it is not disabled
+if not site.wifi24.mesh.disabled(false) then
+	util.add_to_set(wifi_roles['band2g'], 'mesh')
+end
+if not site.wifi5.mesh.disabled(false) then
+	util.add_to_set(wifi_roles['band5g'], 'mesh')
+end
+
+local function add_or_remove_role(roles, role, disabled)
+	if disabled == '1' then
+		util.remove_from_set(roles, role)
+	elseif disabled == '0' then
+		util.add_to_set(roles, role)
+	end
+end
+
+local function migrate_wifi_to_roles(band, roles)
+	wireless.foreach_radio(uci, function(radio)
+		if radio.band ~= band then
+			return
+		end
+		local radio_name = radio['.name']
+		local suffix = radio_name:match('^radio(%d+)$')
+
+		local client_disabled = uci:get('wireless', 'client_radio' .. suffix, 'disabled')
+		local mesh_disabled = uci:get('wireless', 'mesh_radio' .. suffix, 'disabled')
+		local wan_disabled = uci:get('wireless', 'wan_radio' .. suffix, 'disabled')
+		add_or_remove_role(roles, 'client', client_disabled)
+		add_or_remove_role(roles, 'mesh', mesh_disabled)
+		add_or_remove_role(roles, 'private', wan_disabled)
+		if wan_disabled == '0' then
+			local wan_name = 'wan_radio' .. suffix
+			local ssid = uci:get('wireless', wan_name, 'ssid')
+			local key = uci:get('wireless', wan_name, 'key')
+			local encryption = uci:get('wireless', wan_name, 'encryption')
+			local mfp = uci:get('wireless', wan_name, 'ieee80211w')
+			uci:set('gluon', 'wireless', 'private_ssid', ssid)
+			uci:set('gluon', 'wireless', 'private_key', key)
+			uci:set('gluon', 'wireless', 'private_encryption', encryption)
+			uci:set('gluon', 'wireless', 'private_mfp', mfp)
+		end
+	end)
+end
+
+for band_name, band_roles in pairs(wifi_roles) do
+	local raw_band = band_name:gsub("band", "")
+	if not uci:get('gluon', raw_band) then
+		-- remove when upgrades are only possible from v2025.1 onwards
+		migrate_wifi_to_roles(raw_band, band_roles)
+
+		uci:section('gluon', 'wireless_band', raw_band, {
+			role = band_roles,
+		})
+	end
+end
+
+
+local function is_outdoor()
+	return uci:get_bool('gluon', 'wireless', 'outdoor')
+end
+
+-- Fix invalid mesh role configurations for outdoor
+uci:foreach('gluon', 'wireless_band', function(band)
+	local function has_role(role)
+		if band.role == nil then
+			return false
+		end
+
+		return util.contains(band.role, role)
+	end
+
+	if band['.name'].name=='5g' and has_role('mesh') and is_outdoor() then
+		-- remove 'mesh' role
+		util.remove_from_set(band.role, 'mesh')
+		uci:set('gluon', band['.name'], 'role', band.role)
+	end
+end)
+
+uci:save('gluon')

From 5cea4629c07ecc63b789e0237675d76e91a7a1a9 Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Wed, 20 Aug 2025 00:03:54 +0200
Subject: [PATCH 04/12] gluon-core: create wireless config from gluon wireless
 roles section

Delete all wifi-ifaces on reconfigure.
Enable mesh based on the configured wireless role.
Adds logic for client_only_radio detection.

The client_only_radio is extended from the existing draft to theoretically also support 2g multi-band.

If there is an additional radio which does not support the mesh channel, we can configure it on a
different channel. In this case, the other radio on the same band supporting the channel turns off its
client radio and is mesh only.
---
 .../upgrade/320-gluon-client-bridge-wireless  |  22 ++++
 .../luasrc/lib/gluon/upgrade/200-wireless     | 116 +++++++++---------
 .../luasrc/usr/lib/lua/gluon/wireless.lua     |  10 ++
 3 files changed, 87 insertions(+), 61 deletions(-)

diff --git a/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless b/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
index 1fd97301f3..04efdd4be4 100755
--- a/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
+++ b/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
@@ -1,6 +1,7 @@
 #!/usr/bin/lua
 
 local wireless = require 'gluon.wireless'
+local util = require 'gluon.util'
 
 local uci = require('simple-uci').cursor()
 
@@ -103,8 +104,29 @@ local function configure_owe_transition_mode(config, radio_name)
 	uci:set('wireless', name_owe, 'hidden', '1')
 end
 
+local client_only_radio = {band2g=0, band5g=0}
+
+local function count_band(band)
+	client_only_radio[band] = client_only_radio[band] + 1
+end
+
+wireless.foreach_radio(uci, function(radio, _, config)
+	if wireless.supports_channel(radio, config.channel()) then
+		count_band("band" .. config.band)
+	end
+end)
+
 wireless.foreach_radio(uci, function(radio, index, config)
 	local radio_name = radio['.name']
+	-- now we assign radios to roles
+	local band = radio.band
+	local roles = uci:get('gluon', band, 'role')
+
+	-- if there is a radio which does not support the mesh channel, we can configure it on a different channel
+	-- in this case, the other radio on the same band supporting the channel turns off its client radio and is mesh only
+	local disable_client = wireless.supports_channel(radio, config.channel()) and client_only_radio['band' .. band] > 0
+
+	uci:set('wireless', 'client_' .. radio_name, 'disabled', util.contains(roles, 'client') and not disable_client)
 
 	configure_ap(radio, index, config, radio_name)
 
diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
index 904141f619..8ebb173716 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
@@ -3,55 +3,51 @@
 local util = require 'gluon.util'
 local wireless = require 'gluon.wireless'
 local site = require 'gluon.site'
-local sysconfig = require 'gluon.sysconfig'
 local iwinfo = require 'iwinfo'
 
 local uci = require('simple-uci').cursor()
 
--- Initial
-if not sysconfig.gluon_version then
-	uci:delete_all('wireless', 'wifi-iface')
+uci:delete_all('wireless', 'wifi-iface')
 
-	local radio_band_count = {band2g=0, band5g=0}
+local radio_band_count = {band2g=0, band5g=0}
 
-	local function count_band(band)
-		radio_band_count[band] = radio_band_count[band] + 1
+local function count_band(band)
+	radio_band_count[band] = radio_band_count[band] + 1
+end
+
+-- First count all radios with a fixed frequency band.
+-- This is needed to distribute devices which have radios
+-- capable of operating in the 2.4 GHz and 5 GHz band need
+-- to be distributed evenly.
+wireless.foreach_radio(uci, function(radio)
+	local hwmodes = iwinfo.nl80211.hwmodelist(wireless.find_phy(radio))
+	if hwmodes.g and not (hwmodes.a or hwmodes.ac) then
+		-- 2.4 GHz
+		count_band("band2g")
+	elseif (hwmodes.a or hwmodes.ac) and not hwmodes.g then
+		-- 5 GHz
+		count_band("band5g")
 	end
+end)
 
-	-- First count all radios with a fixed frequency band.
-	-- This is needed to distribute devices which have radios
-	-- capable of operating in the 2.4 GHz and 5 GHz band need
-	-- to be distributed evenly.
-	wireless.foreach_radio(uci, function(radio)
-		local hwmodes = iwinfo.nl80211.hwmodelist(wireless.find_phy(radio))
-		if hwmodes.g and not (hwmodes.a or hwmodes.ac) then
-			-- 2.4 GHz
+-- Use the number of all fixed 2.4G GHz and 5 GHz radios to
+-- distribute dualband radios in this step.
+wireless.foreach_radio(uci, function(radio)
+	local radio_name = radio['.name']
+	local hwmodes = iwinfo.nl80211.hwmodelist(wireless.find_phy(radio))
+	if (hwmodes.a or hwmodes.ac) and hwmodes.g then
+		-- Dualband radio
+		if radio_band_count["band2g"] <= radio_band_count["band5g"] then
+			-- Assign radio to 2.4GHz band
 			count_band("band2g")
-		elseif (hwmodes.a or hwmodes.ac) and not hwmodes.g then
-			-- 5 GHz
+			uci:set('wireless', radio_name, 'band', '2g')
+		else
+			-- Assign radio to 5GHz band
 			count_band("band5g")
+			uci:set('wireless', radio_name, 'band', '5g')
 		end
-	end)
-
-	-- Use the number of all fixed 2.4G GHz and 5 GHz radios to
-	-- distribute dualband radios in this step.
-	wireless.foreach_radio(uci, function(radio)
-		local radio_name = radio['.name']
-		local hwmodes = iwinfo.nl80211.hwmodelist(wireless.find_phy(radio))
-		if (hwmodes.a or hwmodes.ac) and hwmodes.g then
-			-- Dualband radio
-			if radio_band_count["band2g"] <= radio_band_count["band5g"] then
-				-- Assign radio to 2.4GHz band
-				count_band("band2g")
-				uci:set('wireless', radio_name, 'band', '2g')
-			else
-				-- Assign radio to 5GHz band
-				count_band("band5g")
-				uci:set('wireless', radio_name, 'band', '5g')
-			end
-		end
-	end)
-end
+	end
+end)
 
 local function is_outdoor()
 	return uci:get_bool('gluon', 'wireless', 'outdoor')
@@ -90,14 +86,6 @@ local function get_htmode(radio)
 	return 'HT20'
 end
 
-local function is_disabled(name)
-	if not uci:get('wireless', name) then
-		return nil
-	end
-
-	return uci:get_bool('wireless', name, 'disabled')
-end
-
 -- Returns the first argument that is not nil; don't call without any non-nil arguments!
 local function first_non_nil(first, ...)
 	if first ~= nil then
@@ -176,7 +164,6 @@ local function configure_mesh_wireless(radio, index, config, disabled)
 	configure_mesh(config.mesh(), radio, index, suffix,
 		first_non_nil(
 			disabled,
-			is_disabled('mesh_' .. radio_name),
 			config.mesh.disabled(false)
 		)
 	)
@@ -225,27 +212,34 @@ wireless.foreach_radio(uci, function(radio, index, config)
 	uci:delete('wireless', radio_name, 'basic_rate')
 
 	local band = radio.band
+	-- now we assign radios to roles
+	local roles = uci:get('gluon', band, 'role')
+
 	if band == '2g' then
 		uci:set('wireless', radio_name, 'legacy_rates', false)
-		configure_mesh_wireless(radio, index, config)
-	elseif (band == '5g') then
-		-- ToDo: Remove in v2024.x
-		local hostapd_options = uci:get_list('wireless', radio_name, 'hostapd_options')
-		util.remove_from_set(hostapd_options, 'country3=0x4f')
-		uci:set_list('wireless', radio_name, 'hostapd_options', hostapd_options)
-
-		if is_outdoor() then
-			-- enforce outdoor channels by filtering the regdom for outdoor channels
-			uci:set('wireless', radio_name, 'country3', '0x4f')
-			configure_mesh_wireless(radio, index, config, true)
-		else
-			uci:delete('wireless', radio_name, 'country3')
-			configure_mesh_wireless(radio, index, config)
+	end
+
+	if roles and util.contains(roles, 'mesh') then
+		if band == '2g' then
+			configure_mesh_wireless(radio, index, config, false)
+		elseif (band == '5g') then
+			if is_outdoor() or not wireless.supports_channel(radio, config.channel()) then
+				-- enforce outdoor channels by filtering the regdom for outdoor channels
+				uci:set('wireless', radio_name, 'country3', '0x4f')
+				configure_mesh_wireless(radio, index, config, true)
+			else
+				uci:delete('wireless', radio_name, 'country3')
+				configure_mesh_wireless(radio, index, config, false)
+			end
 		end
+	else
+		-- disable mesh if role is not configured
+		configure_mesh_wireless(radio, index, config, true)
 	end
 
 	uci:set('wireless', radio_name, 'beacon_int', beacon_interval)
 
+	uci:set('wireless', "wan_" .. radio_name, 'disabled', util.contains(roles, 'private'))
 	fixup_wan(radio, index)
 end)
 
diff --git a/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua b/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua
index 5ded4f5f50..b93ad2566b 100644
--- a/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua
+++ b/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua
@@ -39,6 +39,16 @@ local function get_wlan_mac_from_driver(radio, vif)
 	return addresses[vif+1]
 end
 
+function M.supports_channel(radio, channel)
+	local phy = M.find_phy(radio)
+	for _, chan in ipairs(iwinfo.nl80211.freqlist(phy)) do
+		if channel == chan.channel then
+			return true
+		end
+	end
+	return false
+end
+
 local radio_mac_offsets = {
 	client = 0,
 	mesh = 1,

From da4931ac8bbdf1b75d0d739c559ca19fe7fb6d5e Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Mon, 18 Aug 2025 09:29:31 +0200
Subject: [PATCH 05/12] gluon-core: add hop_penalty support for mesh interfaces
 and fix preserve_channels

do not delete config if preserve_channels is set

allow setting a batadv_hop_penalty for the gluon_mesh wireless interfaces as well
---
 package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
index 8ebb173716..07e7aaab53 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
@@ -7,7 +7,9 @@ local iwinfo = require 'iwinfo'
 
 local uci = require('simple-uci').cursor()
 
-uci:delete_all('wireless', 'wifi-iface')
+if not uci:get_bool('gluon', 'wireless', 'preserve_channels') then
+	uci:delete_all('wireless', 'wifi-iface')
+end
 
 local radio_band_count = {band2g=0, band5g=0}
 
@@ -120,8 +122,11 @@ local function configure_mesh(config, radio, index, suffix, disabled)
 		return
 	end
 
+	local batadv_hop_penalty = uci:get('gluon', radio.band, 'batadv_hop_penalty') or 0
+
 	uci:section('network', 'interface', name, {
 		proto = 'gluon_mesh',
+		hop_penalty = batadv_hop_penalty,
 	})
 
 	uci:section('wireless', 'wifi-iface', name, {

From f7ccf5f1d27957ac93f2fcf5f0774b153ccbffc3 Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Mon, 18 Aug 2025 11:34:32 +0200
Subject: [PATCH 06/12] gluon-client-bridge: correctly enabled/disable client
 and owe based on role

If there is a radio which does not support the mesh channel, we can configure it on a different channel.
In this case, the other radio on the same band supporting the channel turns off its client radio and is mesh only.
---
 .../upgrade/320-gluon-client-bridge-wireless  | 48 ++++++++-----------
 1 file changed, 20 insertions(+), 28 deletions(-)

diff --git a/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless b/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
index 04efdd4be4..76f1fa1553 100755
--- a/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
+++ b/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
@@ -5,13 +5,27 @@ local util = require 'gluon.util'
 
 local uci = require('simple-uci').cursor()
 
+local client_only_radio = {band2g=0, band5g=0}
+
+local function count_band(band)
+	client_only_radio[band] = client_only_radio[band] + 1
+end
 
-local function is_disabled(config, name)
-	if uci:get('wireless', name) then
-		return uci:get_bool('wireless', name, 'disabled')
+wireless.foreach_radio(uci, function(radio, _, config)
+	if not wireless.supports_channel(radio, config.channel()) then
+		count_band("band" .. radio.band)
 	end
+end)
 
-	return config.disabled(false)
+local function is_disabled(radio, config)
+	-- now we assign radios to roles
+	local band = radio.band
+	local roles = uci:get('gluon', band, 'role')
+	-- if there is a radio which does not support the mesh channel, we can configure it on a different channel
+	-- in this case, the other radio on the same band supporting the channel turns off its client radio and is mesh only
+	local disable_client = wireless.supports_channel(radio, config.channel()) and client_only_radio['band' .. band] > 0
+	local client_enabled = roles and util.contains(roles, 'client') and not disable_client
+	return config.ap.disabled(not client_enabled)
 end
 
 local function configure_ap(radio, index, config, radio_name)
@@ -19,7 +33,7 @@ local function configure_ap(radio, index, config, radio_name)
 	local suffix = radio_name:match('^radio(%d+)$')
 
 	local ap = config.ap
-	local disabled = is_disabled(ap, name)
+	local disabled = is_disabled(radio, config)
 
 	uci:delete('wireless', name)
 
@@ -46,7 +60,7 @@ local function configure_owe(radio, index, config, radio_name)
 
 	local ap = config.ap
 
-	local disabled = is_disabled(ap, 'client_' .. radio_name)
+	local disabled = is_disabled(radio, config)
 
 	uci:delete('wireless', name)
 
@@ -104,30 +118,8 @@ local function configure_owe_transition_mode(config, radio_name)
 	uci:set('wireless', name_owe, 'hidden', '1')
 end
 
-local client_only_radio = {band2g=0, band5g=0}
-
-local function count_band(band)
-	client_only_radio[band] = client_only_radio[band] + 1
-end
-
-wireless.foreach_radio(uci, function(radio, _, config)
-	if wireless.supports_channel(radio, config.channel()) then
-		count_band("band" .. config.band)
-	end
-end)
-
 wireless.foreach_radio(uci, function(radio, index, config)
 	local radio_name = radio['.name']
-	-- now we assign radios to roles
-	local band = radio.band
-	local roles = uci:get('gluon', band, 'role')
-
-	-- if there is a radio which does not support the mesh channel, we can configure it on a different channel
-	-- in this case, the other radio on the same band supporting the channel turns off its client radio and is mesh only
-	local disable_client = wireless.supports_channel(radio, config.channel()) and client_only_radio['band' .. band] > 0
-
-	uci:set('wireless', 'client_' .. radio_name, 'disabled', util.contains(roles, 'client') and not disable_client)
-
 	configure_ap(radio, index, config, radio_name)
 
 	configure_owe(radio, index, config, radio_name)

From c65593a8b79eabd707ef81be1566b53bd2854b34 Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Sat, 16 Aug 2025 18:14:22 +0200
Subject: [PATCH 07/12] gluon-private-wifi: add uci wireless config to
 gluon-private-wifi

Recreate private wifi config using upgrade scripts.
This reduces the coupling of gluon-web-private-wifi, which only sets the role of private wifi now.
---
 docs/features/private-wlan.rst                | 24 ++++----
 .../luasrc/lib/gluon/upgrade/200-wireless     | 19 ------
 .../lib/gluon/upgrade/325-gluon-private-wifi  | 49 +++++++++++++---
 .../config-mode/model/admin/privatewifi.lua   | 58 ++++++++-----------
 4 files changed, 75 insertions(+), 75 deletions(-)

diff --git a/docs/features/private-wlan.rst b/docs/features/private-wlan.rst
index d11c9c1cd8..1eefc566b1 100644
--- a/docs/features/private-wlan.rst
+++ b/docs/features/private-wlan.rst
@@ -15,27 +15,23 @@ made optional or completely disabled in the advanced settings tab.
 The private WLAN can be enabled through the config mode if the package ``gluon-web-private-wifi`` is installed.
 You may also enable a private WLAN using the command line::
 
-  RID=0
   SSID="privateWLANname"
   KEY="yoursecret1337password"
 
-  uci set wireless.wan_radio$RID=wifi-iface
-  uci set wireless.wan_radio$RID.device=radio$RID
-  uci set wireless.wan_radio$RID.network=wan
-  uci set wireless.wan_radio$RID.mode=ap
-  uci set wireless.wan_radio$RID.encryption=psk2
-  uci set wireless.wan_radio$RID.ssid="$SSID"
-  uci set wireless.wan_radio$RID.key="$KEY"
-  uci set wireless.wan_radio$RID.disabled=0
-  uci set wireless.wan_radio$RID.macaddr=$(lua -e "print(require('gluon.wireless').get_wlan_mac('wan_radio', $RID))")
-  uci commit
+  uci add_list gluon.2g.role=private
+  uci add_list gluon.5g.role=private
+  uci set gluon.wireless.private_encryption=psk2
+  uci set gluon.wireless.private_ssid="$SSID"
+  uci set gluon.wireless.private_key="$KEY"
+  uci commit gluon
+  gluon-reconfigure
   wifi
 
 Please replace ``$SSID`` by the name of the WLAN and ``$KEY`` by your passphrase (8-63 characters).
-If you have two radios (e.g. 2.4 and 5 GHz) you need to do this for radio0 and radio1.
+If you have two radios (e.g. 2.4 and 5 GHz) you need to do this for 2g and for 5g.
 
 It may also be disabled by running::
 
-  uci set wireless.wan_radio0.disabled=1
-  uci commit
+  uci del_list set gluon.2g.role=private
+  uci commit gluon
   wifi
diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
index 07e7aaab53..b7ac558102 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
@@ -146,22 +146,6 @@ local function configure_mesh(config, radio, index, suffix, disabled)
 	})
 end
 
-local function fixup_wan(radio, index)
-	local radio_name = radio['.name']
-	local name = 'wan_' .. radio_name
-
-	if not uci:get('wireless', name) then
-		return
-	end
-
-	local macaddr = wireless.get_wlan_mac('wan_radio', index, radio)
-	if not macaddr then
-		return
-	end
-
-	uci:set('wireless', name, 'macaddr', macaddr)
-end
-
 local function configure_mesh_wireless(radio, index, config, disabled)
 	local radio_name = radio['.name']
 	local suffix = radio_name:match('^radio(%d+)$')
@@ -243,9 +227,6 @@ wireless.foreach_radio(uci, function(radio, index, config)
 	end
 
 	uci:set('wireless', radio_name, 'beacon_int', beacon_interval)
-
-	uci:set('wireless', "wan_" .. radio_name, 'disabled', util.contains(roles, 'private'))
-	fixup_wan(radio, index)
 end)
 
 
diff --git a/package/gluon-private-wifi/luasrc/lib/gluon/upgrade/325-gluon-private-wifi b/package/gluon-private-wifi/luasrc/lib/gluon/upgrade/325-gluon-private-wifi
index 2485847cd2..efbfcf71f6 100755
--- a/package/gluon-private-wifi/luasrc/lib/gluon/upgrade/325-gluon-private-wifi
+++ b/package/gluon-private-wifi/luasrc/lib/gluon/upgrade/325-gluon-private-wifi
@@ -2,22 +2,55 @@
 
 local uci = require("simple-uci").cursor()
 local wireless = require 'gluon.wireless'
+local util = require 'gluon.util'
 
-wireless.foreach_radio(uci, function(radio)
+local function configure_private_wifi(radio, index)
 	local radio_name = radio['.name']
 	local suffix = radio_name:match('^radio(%d+)$')
 	local name   = "wan_" .. radio_name
 
-	if not uci:get('wireless', name, 'device') then
-		return
+	local encryption = uci:get('gluon', 'wireless', 'private_encryption')
+	local ssid = uci:get('gluon', 'wireless', 'private_ssid')
+	local key = uci:get('gluon', 'wireless', 'private_key')
+	local mfp = wireless.device_supports_mfp(uci) and uci:get('gluon', 'wireless', 'private_mfp')
+
+	-- migrate encryption from Gluon v2023.2.x or older
+	-- remove this in 2027 or on first release supporting only upgrades from >=v2025.1.x
+	encryption = encryption:gsub("psk3", "sae")
+
+	local macaddr = wireless.get_wlan_mac('wan_radio', index, radio)
+	uci:section('wireless', 'wifi-iface', name, {
+		device     = radio_name,
+		network    = 'wan',
+		mode       = 'ap',
+		encryption = encryption,
+		ssid       = ssid,
+		key        = key,
+		macaddr    = macaddr,
+		ifname     = suffix and 'wl-wan' .. suffix,
+		disabled   = false,
+	})
+
+	-- hostapd-mini won't start in case 802.11w is configured
+	if mfp then
+		uci:set('wireless', name, 'ieee80211w', mfp)
+	else
+		uci:delete('wireless', name, 'ieee80211w')
 	end
+end
+
+wireless.foreach_radio(uci, function(radio, index)
+	local radio_name = radio['.name']
+	local name   = "wan_" .. radio_name
 
-	uci:set('wireless', name, 'ifname', suffix and 'wl-wan' .. suffix)
+	local roles = uci:get('gluon', radio.band, 'role')
 
-	-- migrate encryption from Gluon v2023.2.x or older
-	-- remove in 2027 or on first release supporting only upgrades from >=v2025.1.x
-	local encryption = uci:get('wireless', name, 'encryption')
-	uci:set('wireless', name, 'encryption', encryption:gsub("psk3", "sae"))
+	if not roles or not util.contains(roles, 'private') then
+		uci:delete('wireless', name)
+		return
+	end
+
+	configure_private_wifi(radio, index)
 end)
 
 uci:save('wireless')
diff --git a/package/gluon-web-private-wifi/luasrc/lib/gluon/config-mode/model/admin/privatewifi.lua b/package/gluon-web-private-wifi/luasrc/lib/gluon/config-mode/model/admin/privatewifi.lua
index ba649fccee..b15fc6181b 100644
--- a/package/gluon-web-private-wifi/luasrc/lib/gluon/config-mode/model/admin/privatewifi.lua
+++ b/package/gluon-web-private-wifi/luasrc/lib/gluon/config-mode/model/admin/privatewifi.lua
@@ -38,12 +38,12 @@ warning:depends(enabled, true)
 local ssid = s:option(Value, "ssid", translate("Name (SSID)"))
 ssid:depends(enabled, true)
 ssid.datatype = "maxlength(32)"
-ssid.default = uci:get('wireless', primary_iface, "ssid")
+ssid.default = uci:get('gluon', 'wireless', 'private_ssid')
 
 local key = s:option(Value, "key", translate("Key"), translate("8-63 characters"))
 key:depends(enabled, true)
 key.datatype = "wpakey"
-key.default = uci:get('wireless', primary_iface, "key")
+key.default = uci:get('gluon', 'wireless', 'private_key')
 
 local encryption = s:option(ListValue, "encryption", translate("Encryption"))
 encryption:depends(enabled, true)
@@ -52,7 +52,7 @@ if wireless.device_supports_wpa3() then
 	encryption:value("sae-mixed", translate("WPA2 / WPA3"))
 	encryption:value("sae", translate("WPA3"))
 end
-encryption.default = uci:get('wireless', primary_iface, 'encryption') or "psk2"
+encryption.default = uci:get('gluon', 'wireless', 'private_encryption') or "psk2"
 
 local mfp = s:option(ListValue, "mfp", translate("Management Frame Protection"))
 mfp:depends(enabled, true)
@@ -61,41 +61,31 @@ if wireless.device_supports_mfp(uci) then
 	mfp:value("1", translate("Optional"))
 	mfp:value("2", translate("Required"))
 end
-mfp.default = uci:get('wireless', primary_iface, 'ieee80211w') or "0"
+mfp.default = uci:get('gluon', 'wireless', 'private_mfp') or "0"
 
 
 function f:write()
-	wireless.foreach_radio(uci, function(radio, index)
-		local radio_name = radio['.name']
-		local suffix = radio_name:match('^radio(%d+)$')
-		local name   = "wan_" .. radio_name
-
-		if enabled.data then
-			local macaddr = wireless.get_wlan_mac('wan_radio', index, radio)
-
-			uci:section('wireless', 'wifi-iface', name, {
-				device     = radio_name,
-				network    = 'wan',
-				mode       = 'ap',
-				encryption = encryption.data,
-				ssid       = ssid.data,
-				key        = key.data,
-				macaddr    = macaddr,
-				ifname     = suffix and 'wl-wan' .. suffix,
-				disabled   = false,
-			})
-
-			-- hostapd-mini won't start in case 802.11w is configured
-			if wireless.device_supports_mfp(uci) then
-				uci:set('wireless', name, 'ieee80211w', mfp.data)
-			else
-				uci:delete('wireless', name, 'ieee80211w')
-			end
-		else
-			uci:set('wireless', name, "disabled", true)
-		end
-	end)
+	local roles_2g = uci:get_list('gluon', '2g', 'role')
+	local roles_5g = uci:get_list('gluon', '5g', 'role')
+	if enabled.data then
+		-- set new private wifi configuration
+		uci:set('gluon', 'wireless', 'private_ssid', ssid.data)
+		uci:set('gluon', 'wireless', 'private_key', key.data)
+		uci:set('gluon', 'wireless', 'private_encryption', encryption.data)
+		uci:set('gluon', 'wireless', 'private_mfp', mfp.data)
+
+		util.add_to_set(roles_2g, 'private')
+		util.add_to_set(roles_5g, 'private')
+	else
+		util.remove_from_set(roles_2g, 'private')
+		util.remove_from_set(roles_5g, 'private')
+	end
+	uci:set_list('gluon', '2g', 'role', roles_2g)
+	uci:set_list('gluon', '5g', 'role', roles_5g)
 
+	uci:commit('gluon')
+	os.execute('/lib/gluon/upgrade/325-gluon-private-wifi')
+	uci:commit('network')
 	uci:commit('wireless')
 end
 

From fb9ee033817fdd816072f8d6f29e9d41223f1a7c Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Mon, 18 Aug 2025 12:33:20 +0200
Subject: [PATCH 08/12] gluon-web-wifi-config: refactor for gluon roles

Adjust configuration for roles.
Allow setting client and mesh per band instead of per radio.
Add dependency to gluon-client-bridge
---
 package/gluon-web-wifi-config/Makefile        |  2 +-
 .../config-mode/model/admin/wifi-config.lua   | 81 +++++++++----------
 2 files changed, 39 insertions(+), 44 deletions(-)

diff --git a/package/gluon-web-wifi-config/Makefile b/package/gluon-web-wifi-config/Makefile
index e18febad37..07c3faf4a3 100644
--- a/package/gluon-web-wifi-config/Makefile
+++ b/package/gluon-web-wifi-config/Makefile
@@ -5,7 +5,7 @@ PKG_NAME:=gluon-web-wifi-config
 include ../gluon.mk
 
 define Package/gluon-web-wifi-config
-  DEPENDS:=+gluon-web-admin +libiwinfo-lua
+  DEPENDS:=+gluon-web-admin +gluon-client-bridge +libiwinfo-lua
   TITLE:=UI for Wifi Settings
 endef
 
diff --git a/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua b/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
index 92cda8efe0..2ede916813 100644
--- a/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
+++ b/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
@@ -2,7 +2,7 @@ local iwinfo = require 'iwinfo'
 local uci = require("simple-uci").cursor()
 local site = require 'gluon.site'
 local wireless = require 'gluon.wireless'
-
+local util = require 'gluon.util'
 
 local function txpower_list(phy)
 	local list = iwinfo.nl80211.txpwrlist(phy) or { }
@@ -38,15 +38,47 @@ f:section(Section, nil, translate(
 
 local mesh_vifs_5ghz = {}
 
+local function add_or_remove_role(roles, role, disabled)
+	if disabled == true then
+		util.remove_from_set(roles, role)
+	elseif disabled == false then
+		util.add_to_set(roles, role)
+	end
+end
+
+local function vif_option(section, role_name, band, msg)
+	local config_roles = uci:get('gluon', band, 'role') or {}
+	local o = section:option(Flag, band .. '_' .. role_name .. '_enabled', msg)
+	o.default = util.contains(config_roles, role_name)
+
+	function o:write(data)
+		-- this does race for multiple options without additional read before write in o:write
+		local roles = uci:get('gluon', band, 'role') or {}
+		add_or_remove_role(roles, role_name, not data)
+		uci:set('gluon', band, 'role', roles)
+	end
+
+	return o
+end
+
+local bands = {}
 
 uci:foreach('wireless', 'wifi-device', function(config)
 	local radio = config['.name']
 
 	local is_5ghz = false
 	local title
-	if config.band == '2g' then
+	local band = config.band
+
+	if bands and util.contains(bands, band) then
+		return
+	end
+
+	table.insert(bands, band)
+
+	if band == '2g' then
 		title = translate("2.4GHz WLAN")
-	elseif config.band == '5g' then
+	elseif band == '5g' then
 		is_5ghz = true
 		title = translate("5GHz WLAN")
 	else
@@ -55,47 +87,9 @@ uci:foreach('wireless', 'wifi-device', function(config)
 
 	local p = f:section(Section, title)
 
-	local function filter_existing_interfaces(interfaces)
-		local out = {}
-		for _, interface in ipairs(interfaces) do
-			if uci:get('wireless', interface .. '_' .. radio) then
-				table.insert(out, interface)
-			end
-		end
-		return out
-	end
-
-	local function has_active_interfaces(interfaces)
-		for _, interface in ipairs(interfaces) do
-			if not uci:get_bool('wireless', interface .. '_' .. radio, 'disabled') then
-				return true
-			end
-		end
-		return false
-	end
-
-	local function vif_option(name, interfaces, msg)
-		local existing_interfaces = filter_existing_interfaces(interfaces)
-
-		if #existing_interfaces == 0 then
-			return
-		end
-
-		local o = p:option(Flag, radio .. '_' .. name .. '_enabled', msg)
-		o.default = has_active_interfaces(existing_interfaces)
-
-		function o:write(data)
-			for _, interface in ipairs(existing_interfaces) do
-				uci:set('wireless', interface .. '_' .. radio, 'disabled', not data)
-			end
-		end
-
-		return o
-	end
-
-	vif_option('client', {'client', 'owe'}, translate('Enable client network (access point)'))
+	vif_option(p, 'client', band, translate('Enable client network (access point)'))
 
-	local mesh_vif = vif_option('mesh', {'mesh'}, translate("Enable mesh network (802.11s)"))
+	local mesh_vif = vif_option(p, 'mesh', band, translate("Enable mesh network (802.11s)"))
 	if is_5ghz then
 		table.insert(mesh_vifs_5ghz, mesh_vif)
 	end
@@ -187,6 +181,7 @@ end
 function f:write()
 	uci:commit('gluon')
 	os.execute('/lib/gluon/upgrade/200-wireless')
+	os.execute('/lib/gluon/upgrade/320-gluon-client-bridge-wireless')
 	uci:commit('network')
 	uci:commit('wireless')
 end

From 2ca558aa8864c13ce2f6b3dec2cb1f8434ac52c9 Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Tue, 19 Aug 2025 23:18:51 +0200
Subject: [PATCH 09/12] gluon-core: migrate txpower setting and allow setting
 of htmode

add a txpower and htmode per radio section, which can be set using for example

uci set gluon.wireless.htmode_radio0="HE40"

Similarly, txpower_radio0 can be adjusted.
The migration of the existing txpower value takes already place in 022-wireless-roles
---
 .../lib/gluon/upgrade/022-wireless-roles      | 13 +++
 .../luasrc/lib/gluon/upgrade/200-wireless     | 18 +++-
 .../config-mode/model/admin/wifi-config.lua   | 91 +++++++++----------
 3 files changed, 67 insertions(+), 55 deletions(-)

diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles b/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles
index c62cd71f19..dfe31c8347 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles
@@ -52,6 +52,19 @@ local function migrate_wifi_to_roles(band, roles)
 			uci:set('gluon', 'wireless', 'private_encryption', encryption)
 			uci:set('gluon', 'wireless', 'private_mfp', mfp)
 		end
+
+		local txpower = uci:get('wireless', radio_name, 'txpower')
+		if txpower ~= nil then
+			uci:set('gluon', 'wireless', 'txpower_' .. radio['.name'], txpower)
+		end
+
+		if radio.band == '5g' then
+			local outdoor_htmode = uci:get('gluon', 'wireless', 'outdoor_' .. radio['.name'] .. '_htmode')
+			if outdoor_htmode ~= nil then
+				uci:set('gluon', 'wireless', 'htmode_' .. radio['.name'], outdoor_htmode)
+				uci:delete('gluon', 'wireless', 'outdoor_' .. radio['.name'] .. '_htmode')
+			end
+		end
 	end)
 end
 
diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
index b7ac558102..55e11e05ef 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
@@ -64,16 +64,19 @@ local function get_channel(radio, config)
 	return config.channel()
 end
 
+local function get_txpower(radio)
+	return uci:get('gluon', 'wireless', 'txpower_' .. radio['.name'])
+end
+
 local function get_htmode(radio)
 	if wireless.preserve_channels(uci) then
 		return radio.htmode
 	end
 
-	if radio.band == '5g' and is_outdoor() then
-		local outdoor_htmode = uci:get('gluon', 'wireless', 'outdoor_' .. radio['.name'] .. '_htmode')
-		if outdoor_htmode ~= nil then
-			return outdoor_htmode
-		end
+
+	local htmode = uci:get('gluon', 'wireless', 'htmode_' .. radio['.name'])
+	if htmode ~= nil then
+		return htmode
 	end
 
 	local phy = wireless.find_phy(radio)
@@ -187,6 +190,11 @@ wireless.foreach_radio(uci, function(radio, index, config)
 		return
 	end
 
+	local txpower = get_txpower(radio)
+	if txpower ~= nil then
+		uci:set('wireless', radio_name, 'txpower', txpower)
+	end
+
 	local htmode = get_htmode(radio)
 	local beacon_interval = config.beacon_interval()
 
diff --git a/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua b/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
index 2ede916813..087e9251c9 100644
--- a/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
+++ b/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
@@ -70,30 +70,32 @@ uci:foreach('wireless', 'wifi-device', function(config)
 	local title
 	local band = config.band
 
-	if bands and util.contains(bands, band) then
-		return
-	end
-
-	table.insert(bands, band)
-
-	if band == '2g' then
-		title = translate("2.4GHz WLAN")
-	elseif band == '5g' then
-		is_5ghz = true
-		title = translate("5GHz WLAN")
+	local p
+	-- store section to add txpower and htmode to
+	if bands and bands[band] then
+		p = bands[band]
 	else
-		return
-	end
+		if band == '2g' then
+			title = translate("2.4GHz WLAN")
+		elseif band == '5g' then
+			is_5ghz = true
+			title = translate("5GHz WLAN")
+		else
+			return
+		end
 
-	local p = f:section(Section, title)
+		p = f:section(Section, title)
 
-	vif_option(p, 'client', band, translate('Enable client network (access point)'))
+		vif_option(p, 'client', band, translate('Enable client network (access point)'))
 
-	local mesh_vif = vif_option(p, 'mesh', band, translate("Enable mesh network (802.11s)"))
-	if is_5ghz then
-		table.insert(mesh_vifs_5ghz, mesh_vif)
+		local mesh_vif = vif_option(p, 'mesh', band, translate("Enable mesh network (802.11s)"))
+		if is_5ghz then
+			table.insert(mesh_vifs_5ghz, mesh_vif)
+		end
+		bands[band] = p
 	end
 
+	-- txpowerlist
 	local phy = wireless.find_phy(config)
 	if not phy then
 		return
@@ -104,8 +106,8 @@ uci:foreach('wireless', 'wifi-device', function(config)
 		return
 	end
 
-	local tp = p:option(ListValue, radio .. '_txpower', translate("Transmission power"))
-	tp.default = uci:get('wireless', radio, 'txpower') or 'default'
+	local tp = p:option(ListValue, 'txpower_' .. radio, translate("Transmission power").. ' (' .. radio .. ')')
+	tp.default = uci:get('gluon', 'wireless', 'txpower_' .. radio) or 'default'
 
 	tp:value('default', translate("(default)"))
 
@@ -119,7 +121,25 @@ uci:foreach('wireless', 'wifi-device', function(config)
 		if data == 'default' then
 			data = nil
 		end
-		uci:set('wireless', radio, 'txpower', data)
+		uci:set('gluon', 'wireless', 'txpower_' .. radio, data)
+	end
+
+	-- htmode
+	local ht = p:option(ListValue, 'htmode_' .. radio, translate('HT Mode') .. ' (' .. radio .. ')')
+	ht.default = uci:get('gluon', 'wireless', 'htmode_' .. radio) or 'default'
+	ht:value('default', translate("(default)"))
+
+	for mode, available in pairs(iwinfo.nl80211.htmodelist(phy)) do
+		if available then
+			ht:value(mode, mode)
+		end
+	end
+
+	function ht:write(data)
+		if data == 'default' then
+			data = nil
+		end
+		uci:set('gluon', 'wireless', 'htmode_' .. radio, data)
 	end
 end)
 
@@ -146,35 +166,6 @@ if wireless.device_uses_5ghz(uci) and not wireless.preserve_channels(uci) then
 	function outdoor:write(data)
 		uci:set('gluon', 'wireless', 'outdoor', data)
 	end
-
-	uci:foreach('wireless', 'wifi-device', function(config)
-		local radio = config['.name']
-		local band = uci:get('wireless', radio, 'band')
-
-		if band ~= '5g' then
-			return
-		end
-
-		local phy = wireless.find_phy(config)
-
-		local ht = r:option(ListValue, 'outdoor_htmode', translate('HT Mode') .. ' (' .. radio .. ')')
-		ht:depends(outdoor, true)
-		ht.default = uci:get('gluon', 'wireless', 'outdoor_' .. radio .. '_htmode') or 'default'
-
-		ht:value('default', translate("(default)"))
-		for mode, available in pairs(iwinfo.nl80211.htmodelist(phy)) do
-			if available then
-				ht:value(mode, mode)
-			end
-		end
-
-		function ht:write(data)
-			if data == 'default' then
-				data = nil
-			end
-			uci:set('gluon', 'wireless', 'outdoor_' .. radio .. '_htmode', data)
-		end
-	end)
 end
 
 

From 234ae5f2a89247446289379ac82da1e60e4140a8 Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Sat, 6 Sep 2025 02:51:52 +0200
Subject: [PATCH 10/12] add vero w6m  to supported devices

---
 docs/user/supported_devices.rst | 4 ++++
 targets/mediatek-filogic        | 7 +++++++
 2 files changed, 11 insertions(+)

diff --git a/docs/user/supported_devices.rst b/docs/user/supported_devices.rst
index 67a7b847ed..fb0a6822aa 100644
--- a/docs/user/supported_devices.rst
+++ b/docs/user/supported_devices.rst
@@ -313,6 +313,10 @@ lantiq-xway
 mediatek-filogic
 ----------------
 
+* Acer
+
+  - Vero-W6M
+
 * ASUS
 
   - RT-AX52
diff --git a/targets/mediatek-filogic b/targets/mediatek-filogic
index 2a399c500b..818bb01114 100644
--- a/targets/mediatek-filogic
+++ b/targets/mediatek-filogic
@@ -1,6 +1,13 @@
 config('TARGET_ROOTFS_INITRAMFS', true)
 config('TARGET_INITRAMFS_COMPRESSION_LZMA', true)
 
+-- Acer
+
+device('acer-vero-w6m', 'acer_vero-w6m', {
+	factory = false,
+})
+
+
 -- ASUS
 
 device('asus-rt-ax52', 'asus_rt-ax52', {

From f96d9bbfb78e8abe11498497b173a1c21beb7d87 Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Mon, 18 Aug 2025 10:50:37 +0200
Subject: [PATCH 11/12] gluon-core: add basic support for 6GHz wifi

---
 .../lib/gluon/upgrade/320-gluon-client-bridge-wireless |  2 +-
 .../luasrc/lib/gluon/upgrade/022-wireless-roles        |  4 ++++
 .../gluon-core/luasrc/lib/gluon/upgrade/200-wireless   | 10 +++++++++-
 .../lib/gluon/config-mode/model/admin/privatewifi.lua  |  5 +++++
 .../lib/gluon/config-mode/model/admin/wifi-config.lua  |  2 ++
 5 files changed, 21 insertions(+), 2 deletions(-)

diff --git a/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless b/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
index 76f1fa1553..315f2f5d12 100755
--- a/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
+++ b/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
@@ -5,7 +5,7 @@ local util = require 'gluon.util'
 
 local uci = require('simple-uci').cursor()
 
-local client_only_radio = {band2g=0, band5g=0}
+local client_only_radio = {band2g=0, band5g=0, band6g=0}
 
 local function count_band(band)
 	client_only_radio[band] = client_only_radio[band] + 1
diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles b/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles
index dfe31c8347..9decf6570e 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/022-wireless-roles
@@ -9,6 +9,7 @@ local wireless = require 'gluon.wireless'
 local wifi_roles = {
 	band2g = {'client'},
 	band5g = {'client'},
+	band6g = {'client'},
 }
 
 -- enable mesh if it is not disabled
@@ -18,6 +19,9 @@ end
 if not site.wifi5.mesh.disabled(false) then
 	util.add_to_set(wifi_roles['band5g'], 'mesh')
 end
+if not site.wifi6.mesh.disabled(false) then
+	util.add_to_set(wifi_roles['band6g'], 'mesh')
+end
 
 local function add_or_remove_role(roles, role, disabled)
 	if disabled == '1' then
diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
index 55e11e05ef..423695467c 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
@@ -11,7 +11,7 @@ if not uci:get_bool('gluon', 'wireless', 'preserve_channels') then
 	uci:delete_all('wireless', 'wifi-iface')
 end
 
-local radio_band_count = {band2g=0, band5g=0}
+local radio_band_count = {band2g=0, band5g=0, band6g=0}
 
 local function count_band(band)
 	radio_band_count[band] = radio_band_count[band] + 1
@@ -29,6 +29,10 @@ wireless.foreach_radio(uci, function(radio)
 	elseif (hwmodes.a or hwmodes.ac) and not hwmodes.g then
 		-- 5 GHz
 		count_band("band5g")
+	elseif hwmodes.ax and not hwmodes.ac then
+		-- TODO validate
+		-- 6 GHz
+		count_band("band6g")
 	end
 end)
 
@@ -61,6 +65,10 @@ local function get_channel(radio, config)
 		return 'auto'
 	end
 
+	if radio.band == '6g' then
+		return 'auto'
+	end
+
 	return config.channel()
 end
 
diff --git a/package/gluon-web-private-wifi/luasrc/lib/gluon/config-mode/model/admin/privatewifi.lua b/package/gluon-web-private-wifi/luasrc/lib/gluon/config-mode/model/admin/privatewifi.lua
index b15fc6181b..2336550582 100644
--- a/package/gluon-web-private-wifi/luasrc/lib/gluon/config-mode/model/admin/privatewifi.lua
+++ b/package/gluon-web-private-wifi/luasrc/lib/gluon/config-mode/model/admin/privatewifi.lua
@@ -67,8 +67,10 @@ mfp.default = uci:get('gluon', 'wireless', 'private_mfp') or "0"
 function f:write()
 	local roles_2g = uci:get_list('gluon', '2g', 'role')
 	local roles_5g = uci:get_list('gluon', '5g', 'role')
+	local roles_6g = uci:get_list('gluon', '6g', 'role')
 	if enabled.data then
 		-- set new private wifi configuration
+		-- TODO should this be set everytime? To reset this with empty values?
 		uci:set('gluon', 'wireless', 'private_ssid', ssid.data)
 		uci:set('gluon', 'wireless', 'private_key', key.data)
 		uci:set('gluon', 'wireless', 'private_encryption', encryption.data)
@@ -76,12 +78,15 @@ function f:write()
 
 		util.add_to_set(roles_2g, 'private')
 		util.add_to_set(roles_5g, 'private')
+		util.add_to_set(roles_6g, 'private')
 	else
 		util.remove_from_set(roles_2g, 'private')
 		util.remove_from_set(roles_5g, 'private')
+		util.remove_from_set(roles_6g, 'private')
 	end
 	uci:set_list('gluon', '2g', 'role', roles_2g)
 	uci:set_list('gluon', '5g', 'role', roles_5g)
+	uci:set_list('gluon', '6g', 'role', roles_6g)
 
 	uci:commit('gluon')
 	os.execute('/lib/gluon/upgrade/325-gluon-private-wifi')
diff --git a/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua b/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
index 087e9251c9..20f10d1380 100644
--- a/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
+++ b/package/gluon-web-wifi-config/luasrc/lib/gluon/config-mode/model/admin/wifi-config.lua
@@ -77,6 +77,8 @@ uci:foreach('wireless', 'wifi-device', function(config)
 	else
 		if band == '2g' then
 			title = translate("2.4GHz WLAN")
+		elseif band == '6g' then
+			title = translate("6GHz WLAN")
 		elseif band == '5g' then
 			is_5ghz = true
 			title = translate("5GHz WLAN")

From f8e06d287b1c35e939d274b82421061444a05b73 Mon Sep 17 00:00:00 2001
From: Florian Maurer <f.maurer@outlook.de>
Date: Sat, 6 Sep 2025 02:46:16 +0200
Subject: [PATCH 12/12] update 6ghz

---
 package/gluon-client-bridge/check_site.lua            |  2 +-
 .../gluon/upgrade/320-gluon-client-bridge-wireless    | 11 +++++++++--
 package/gluon-core/check_site.lua                     |  8 +++++++-
 .../gluon-core/luasrc/lib/gluon/upgrade/200-wireless  |  6 +-----
 .../gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua  |  2 ++
 .../luasrc/lib/gluon/upgrade/205-wireless-mesh-sae    |  2 +-
 6 files changed, 21 insertions(+), 10 deletions(-)

diff --git a/package/gluon-client-bridge/check_site.lua b/package/gluon-client-bridge/check_site.lua
index b077f87bb5..8cbf2ffca0 100644
--- a/package/gluon-client-bridge/check_site.lua
+++ b/package/gluon-client-bridge/check_site.lua
@@ -7,7 +7,7 @@ end
 need_string_match(in_domain({'next_node', 'ip6'}), '^[%x:]+$', false)
 
 
-for _, config in ipairs({'wifi24', 'wifi5'}) do
+for _, config in ipairs({'wifi24', 'wifi5', 'wifi6'}) do
 	if need_table({config, 'ap'}, nil, false) then
 		need_boolean({config, 'ap', 'disabled'}, false)
 		if need_boolean({config, 'ap', 'owe_transition_mode'}, false) then
diff --git a/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless b/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
index 315f2f5d12..85424c3c55 100755
--- a/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
+++ b/package/gluon-client-bridge/luasrc/lib/gluon/upgrade/320-gluon-client-bridge-wireless
@@ -120,10 +120,17 @@ end
 
 wireless.foreach_radio(uci, function(radio, index, config)
 	local radio_name = radio['.name']
-	configure_ap(radio, index, config, radio_name)
+	-- 6g does not support unencrypted client
+	if radio.band ~= '6g' then
+		configure_ap(radio, index, config, radio_name)
+	end
 
 	configure_owe(radio, index, config, radio_name)
-	configure_owe_transition_mode(config, radio_name)
+
+	-- 6g does not support unencrypted client
+	if radio.band ~= '6g' then
+		configure_owe_transition_mode(config, radio_name)
+	end
 end)
 
 uci:save('wireless')
diff --git a/package/gluon-core/check_site.lua b/package/gluon-core/check_site.lua
index cf72f05fc8..c72dab3774 100644
--- a/package/gluon-core/check_site.lua
+++ b/package/gluon-core/check_site.lua
@@ -31,7 +31,7 @@ need_string_match(in_domain({'prefix6'}), '^[%x:]+/64$')
 need_string_array_match(in_domain({'extra_prefixes6'}), '^[%x:]+/%d+$', false)
 
 local supported_rates = {6000, 9000, 12000, 18000, 24000, 36000, 48000, 54000}
-for _, config in ipairs({'wifi24', 'wifi5'}) do
+for _, config in ipairs({'wifi24', 'wifi5', 'wifi6'}) do
 	if need_table({config}, nil, false) then
 		need_string(in_site({'regdom'})) -- regdom is only required when wifi24 or wifi5 is configured
 		need_number({config, 'beacon_interval'}, false)
@@ -48,6 +48,12 @@ for _, config in ipairs({'wifi24', 'wifi5'}) do
 			need_one_of({config, 'channel'}, channels)
 			need_chanlist({config, 'outdoor_chanlist'}, channels, false)
 			need_one_of({config, 'outdoors'}, {true, false, 'preset'}, false)
+		elseif config == 'wifi6' then
+			local channels = {
+				1, 5, 9, 13, 17, 21, 25, 29, 33, 37, 41, 45, 49, 53, 57, 61,
+				65, 69, 73, 77, 81, 85, 89, 93 }
+			need_one_of({config, 'channel'}, channels)
+			need_string_match(in_domain({config, 'ap', 'owe_ssid'}), '^' .. ('.?'):rep(32) .. '$', false)
 		end
 
 		obsolete({config, 'supported_rates'}, '802.11b rates are disabled by default.')
diff --git a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
index 423695467c..cb88463b7e 100755
--- a/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
+++ b/package/gluon-core/luasrc/lib/gluon/upgrade/200-wireless
@@ -65,10 +65,6 @@ local function get_channel(radio, config)
 		return 'auto'
 	end
 
-	if radio.band == '6g' then
-		return 'auto'
-	end
-
 	return config.channel()
 end
 
@@ -225,7 +221,7 @@ wireless.foreach_radio(uci, function(radio, index, config)
 	end
 
 	if roles and util.contains(roles, 'mesh') then
-		if band == '2g' then
+		if band == '2g' or band == '6g' then
 			configure_mesh_wireless(radio, index, config, false)
 		elseif (band == '5g') then
 			if is_outdoor() or not wireless.supports_channel(radio, config.channel()) then
diff --git a/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua b/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua
index b93ad2566b..d596c8650e 100644
--- a/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua
+++ b/package/gluon-core/luasrc/usr/lib/lua/gluon/wireless.lua
@@ -89,6 +89,8 @@ function M.foreach_radio(uci, f)
 			f(radio, index-1, site.wifi24)
 		elseif band == '5g' then
 			f(radio, index-1, site.wifi5)
+		elseif band == '6g' then
+			f(radio, index-1, site.wifi6)
 		end
 	end
 end
diff --git a/package/gluon-mesh-wireless-sae/luasrc/lib/gluon/upgrade/205-wireless-mesh-sae b/package/gluon-mesh-wireless-sae/luasrc/lib/gluon/upgrade/205-wireless-mesh-sae
index a229896694..99f69f3ca1 100755
--- a/package/gluon-mesh-wireless-sae/luasrc/lib/gluon/upgrade/205-wireless-mesh-sae
+++ b/package/gluon-mesh-wireless-sae/luasrc/lib/gluon/upgrade/205-wireless-mesh-sae
@@ -20,7 +20,7 @@ wireless.foreach_radio(uci, function(radio)
 		uci:delete('wireless', vif, 'encryption')
 		uci:delete('wireless', vif, 'key')
 
-		if enable then
+		if enable or radio.band == '6g' then
 			configure_sae(vif)
 		end
 	end
